File: src/main/guitest/amplide-demo/ampl/amplapi/examples/java/EfficientFrontier.java


import java.io.IOException;

import com.ampl.AMPL;
import com.ampl.Objective;
import com.ampl.Parameter;
import com.ampl.Variable;

public class EfficientFrontier {

	public static void main(String[] args) {
		// Number of steps of the efficient frontier
		int steps = 10;
		// Use the provided path or the default one
		String modelDirectory = args.length > 0 ? args[0]
				: "../models";


		AMPL ampl = null;
		// Outer try-catch-finally block, to be sure of releasing the AMPL
		// object when done
		try {
			// Create AMPL object
			ampl= new AMPL();
			
			ampl.setBoolOption("reset_initial_guesses", true);
			ampl.setBoolOption("send_statuses", false);
			ampl.setOption("solver", "cplex");

			// Load the AMPL model from file
			ampl.read(modelDirectory + "/qpmv.mod");
			ampl.read(modelDirectory + "/qpmvbit.run");

			// Set tables directory (parameter used in the script above)
			ampl.getParameter("data_dir").set(modelDirectory);
			// Read tables
			ampl.readTable("assetstable");
			ampl.readTable("astrets");
			
			Variable portfolioReturn = ampl.getVariable("portret");
			Parameter averageReturn = ampl.getParameter("averret");
			Parameter targetReturn = ampl.getParameter("targetret");
			Objective deviation = ampl.getObjective("cst");

			
			ampl.eval("let stockopall:={};let stockrun:=stockall;");
			
			
			// Relax the integrality
			ampl.setBoolOption("relax_integrality", true);
			// Solve the problem
			ampl.solve();
			// Calibrate the efficient frontier range
			double minret = portfolioReturn.value();
			double maxret = findMax(averageReturn.getValues()
					.getColumnAsDoubles("val"));
			double stepsize = (maxret - minret) / steps;

			double[] returns = new double[steps];
			double[] deviations = new double[steps];

			for (int i = 0; i < steps; i++) {
				System.out.format("Solving for return = %f%n", maxret - (i - 1)
						* stepsize);
				// Set target return to the desired point
				targetReturn.setValues(maxret - (i - 1) * stepsize);
				ampl.eval("let stockopall:={};let stockrun:=stockall;");
				// Relax integrality
				ampl.setBoolOption("relax_integrality", true);
				ampl.solve();
				System.out.format("QP result = %f %n", deviation.value());
				// Adjust Included Stocks
				ampl.eval("let stockrun2:={i in stockrun:weights[i]>0};");
				ampl.eval("let stockrun:=stockrun2;");
				ampl.eval("let stockopall:={i in stockrun:weights[i]>0.5};");
				// Set integrality back
				ampl.setBoolOption("relax_integrality", false);
				ampl.solve();
				System.out.format("QMIP result = %f%n", deviation.value());
				// Store data of corrent frontier point
				returns[i] = maxret - (i - 1) * stepsize;
				deviations[i] = deviation.value();
			}

			// Display efficient frontier points
			System.out.format("%-8s  %-8s%n", "RETURN", "DEVIATION");
			for (int i = 0; i < returns.length; i++)
				System.out.format("%-6f  %-6f%n", returns[i], deviations[i]);
		} catch (IOException e) {
			throw new RuntimeException(e);
		} finally {
			ampl.close();
		}
	}

	private static double findMax(double[] array) {
		double max = Double.NEGATIVE_INFINITY;
		for (double d : array)
			if (d > max)
				max = d;
		return max;

	}
}

File: src/main/guitest/amplide-demo/ampl/amplapi/examples/matlab/efficientFrontier.m
% Efficient frontier
function res = efficientFrontier
% PARAMETERS: cutoff points,
% for stocks which seemed very profitable at the relaxed execution
lowcutoff = 0.04;
highcutoff = 0.1;
steps = 30;


% Initialisation
basef = fileparts(which('efficientFrontier'));
addpath(fullfile(basef, '../../matlab'));
ampl = AMPL;
modeldirectory = fullfile(basef, '..', 'models');
modelFile = fullfile(modeldirectory, 'qpmv.mod');
scriptFile = fullfile(modeldirectory, 'qpmvbit.run');

% Load from file the ampl model
ampl.read(modelFile)
% Load the data tables declaration
ampl.read(scriptFile);
% Set the data file location (used in the script)
ampl.getParameter('data_dir').set(modeldirectory);



% Load the data tables
ampl.readTable('assetstable');
ampl.readTable('astrets');
			

% Gets the needed entities
portfolioReturn = ampl.getVariable('portret');
averageReturn = ampl.getParameter('averret');
targetReturn = ampl.getParameter('targetret');
hold = ampl.getVariable('weights');
deviation = ampl.getObjective('cst');

ampl.setOption('solver' ,'cplex');
ampl.eval('let stockopall:={};let stockrun:=stockall;');
% Relax the integerality
ampl.setBoolOption('relax_integrality', true);


% Solve the problem
ampl.solve()
% Calibrate the efficient frontier range
minret = portfolioReturn.getValues().getColumnAsDoubles('val');
maxret = max(averageReturn.getValues().getColumnAsDoubles('val'));
stepsize = (maxret-minret)/steps;
% Create the vectors to store the returns and the deviations
returns = zeros(steps, 1);
deviations = zeros(steps, 1);

% applies a small heuristic (similar to trackingheuristic) to speed up the
% solution of the efficient frontier problems at each step
for i=1:steps  
    currentReturn =  maxret - (i-1)*stepsize;
    fprintf('Solving for return = %f\n',currentReturn );
    targetReturn.setValues(currentReturn);
    ampl.eval('let stockopall:={};let stockrun:=stockall;');
    ampl.eval('options relax_integrality 1;');
    ampl.solve();
    fprintf('QP result = %f ', deviation.value());
    ampl.eval('let stockrun2:={i in stockrun:weights[i]>0};');
	ampl.eval('let stockrun:=stockrun2;let stockopall:={i in stockrun:weights[i]>0.5};');
    ampl.setBoolOption('relax_integrality', false);
    ampl.solve();
    fprintf(' QMIP result = %f\n ', deviation.value());
    returns(i) = maxret - (i-1)*stepsize;
    deviations(i) = deviation.value();
end


%plot(returns, deviations)

res = returns

% Close the AMPL object
ampl.close();